string.c

```diff
@@ -11267,8 +11267,15 @@ rb_str_crypt(VALUE str, VALUE salt)
         CRYPT_END();
         rb_syserr_fail(err, "crypt");
     }
-    result = rb_str_new_cstr(res);
+
+    size_t res_size = strlen(res)+1;
+    char *dup = malloc(res_size); // need to hold onto lock while duplicating a potentially static buffer
```

Comment:
The encrypted result would be small enough in `alloca`.
And this copy should not be needed if `crypt_r` is available.